smithy-typescript-codegen/src/main/java/software/amazon/smithy/typescript/codegen/integration/HttpBindingProtocolGenerator.java

```diff
@@ -478,44 +460,99 @@ protected abstract void serializeDocument(
      * @param context The generation context.
      * @param shape The structure shape being generated.
      */
-    protected abstract void serializeDocumentStructure(
-            GenerationContext context,
-            StructureShape shape
-    );
+    protected abstract void serializeDocumentStructure(GenerationContext context, StructureShape shape);
 
     /**
      * Writes the code needed to serialize a union in the document of a request.
      *
      * @param context The generation context.
      * @param shape The union shape being generated.
      */
-    protected abstract void serializeDocumentUnion(
-            GenerationContext context,
-            UnionShape shape
-    );
+    protected abstract void serializeDocumentUnion(GenerationContext context, UnionShape shape);
 
     /**
      * Writes the code needed to serialize a collection in the document of a request.
      *
      * @param context The generation context.
      * @param shape The collection shape being generated.
      */
-    protected abstract void serializeDocumentCollection(
-            GenerationContext context,
-            CollectionShape shape
-    );
+    protected abstract void serializeDocumentCollection(GenerationContext context, CollectionShape shape);
 
     /**
      * Writes the code needed to serialize a map in the document of a request.
      *
      * @param context The generation context.
      * @param shape The map shape being generated.
      */
-    protected abstract void serializeDocumentMap(
+    protected abstract void serializeDocumentMap(GenerationContext context, MapShape shape);
+
+    /**
+     * Writes the code that dispatches to individual error deserializers. This should
+     * also generate a default of an UnknownError using the passed namespace.
+     *
+     * @param context The generation context.
+     * @param errors The list of errors to dispatch to.
+     * @param unknownErrorNamespace The namespace to use for the default error.
+     */
+    protected abstract void writeErrorDeserializationDispatcher(
             GenerationContext context,
-            MapShape shape
+            List<ShapeId> errors,
+            String unknownErrorNamespace
     );
 
+    /**
+     * Writes the code needed to deserialize an error structure when received.
+     *
+     * @param context The generation context.
+     * @param error The error shape being generated.
+     */
+    protected abstract void deserializeError(GenerationContext context, StructureShape error);
+
+    /**
+     * Writes the code needed to deserialize the output document of a response.
+     *
+     * @param context The generation context.
+     * @param operation The operation being generated.
+     * @param documentBindings The bindings to read from the document.
+     */
+    protected abstract void deserializeDocument(
```

Comment:
What does this actually do? How does someone implement this correctly? What's expected to be generated? If this is the primary method that gets called first, ideally it's the first deserializer of it's kind in the code.